### DISCUSSION TITLE
v20.1.0

### DISCUSSION THREAD

**Discussion by ghost** [uNetworking/uWebSockets.js#632] [General]
# Node.js 17 and BoringSSL
* Adds support for Node.js 17, removes support for Node.js 12.
* Updates to latest uWS.
* Binaries are now built against BoringSSL, not against whatever OpenSSL version/fork Node.js ships with. This improves performance, cuts dependencies on the Node.js camp and improves binary compatibility. **Please test SSL and report any issues.**

<hr /><em>This discussion was created from the release <a href='https://github.com/uNetworking/uWebSockets.js/releases/tag/v20.1.0'>v20.1.0</a>.</em>

**Comment by davidmurdoch** [comment]
Dropping Node 12 seems premature as it is still under LTS until 2022-04-30. I'm guessing there is a technical reason related to BoringSSL for dropping support? If so, do you think it is something that could be fixed if someone did the work and issued a PR for it?

**Reply by ghost** [reply]
There's not much work needed to bring back 12. I deleted it because it was the oldest one. If Node.js follows a release schedule where there always are 3 LTS releases and 1 Current then it could make sense to always support those 4 versions and delete everything else.

**Reply by ghost** [reply]
I don't see why we just delete everything but the last 3 LTS versions - then we only change versions every year and hold 3 binaries per platform, always, which is a nice compromise. No sane company would every run non-LTS anyways. And non-LTS versions are only developed and supported for 8 months.

**Reply by davidmurdoch** [reply]
Out primary use case is developer tooling, so we don't have much say as to what versions of Node our software runs on. We actually just dropped Node 10 support last month. We'll likely drop Node 12 support shortly after the final LTS release in April.

I think shipping binaries for 12, 14, 16, and 17 makes sense _right now_. Then once 18 is out and 12 reaches EOL, shipping binaries for only 14, 16, and 18 would make sense, dropping v17, as hopefully no one relies on Node's odd number releases and expects long term support.

In any case, at Truffle we will need to figure out a solution for Node 12 -- which might just mean we use our pure JS fallback for those devs that are currently stuck (for whatever reason).

**Reply by ghost** [reply]
That's too complex. I'll just go for the 3 LTS versions. That makes things constant - always 3 binaries per platform and always LTS which is what companies use.

**Reply by davidmurdoch** [reply]
Many companies are very slow at updating node versions (the company I work for included!). Google Firebase/Cloud Functions actually didn't make Node 12 ready for "General Availability" until after Node 14 was released! They don't lag too far behind anymore, but for a while we had to support node 10 because of that.

I'm not trying to argue, btw, I'd love to be able to drop support for old node versions in the software I maintain! And fewer binaries in uWS will shrink our own packages, which will be a great thing! I'm only suggesting that you might get some users opening issues about version support every once in a while.

Does the 3 LTS versions plan mean you are planning on adding Node 12 back in and dropping the Node 17 binaries (shipping binaries for Node 12, 14, and 16)? (EDIT: nevermind this question. I just saw the LTS issue you created for it https://github.com/uNetworking/uWebSockets.js/issues/636)

**Reply by Aschen** [reply]
Just my 2 cents, the odd version of Node.js (13, 15, 17, etc) are not considered as LTS and they only have 6 months of official support
